pkg/controller/gc_controller.go

```diff
@@ -84,9 +86,63 @@ func NewGCController(
 		restoreClient:       restoreClient,
 		logger:              logger,
 	}
-}
 
-var _ Interface = &gcController{}
+	backupInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			UpdateFunc: func(old interface{}, new interface{}) {
+				var (
+					backup = new.(*api.Backup)
+					log    = logger.WithField("backup", kube.NamespaceAndName(backup))
+					idx    = -1
+				)
+
+				// we're only interested in backups that have a deletionTimestamp and at
+				// least one finalizer.
+				if backup.DeletionTimestamp == nil || len(backup.Finalizers) == 0 {
+					return
+				}
+				log.Debugf("Backup has finalizers %s", backup.Finalizers)
+
+				for x, finalizer := range backup.Finalizers {
+					if finalizer == gcFinalizer {
+						idx = x
+						break
+					}
+				}
+
+				// GC finalizer not found
+				if idx == -1 {
+					return
+				}
+
+				log.Infof("Garbage-collecting backup")
+				if err := c.garbageCollect(backup, log); err != nil {
+					log.WithError(err).Error("Error deleting backup's related objects")
+					return
+				}
+
+				backup.Finalizers = append(backup.Finalizers[0:idx], backup.Finalizers[idx+1:]...)
+
+				upd, err := backupClient.Backups(api.DefaultNamespace).Update(backup)
+				if err != nil {
+					log.WithError(errors.WithStack(err)).Error("Error updating backup")
+					return
+				}
+
+				// no more finalizers: we don't need to issue another delete
```

Comment:
Shouldn't this be "There are other finalizers, so we won't issue a delete as other controllers still need to process" along with a len >= 1 check?

Reply:
Not 100% sure on how this is supposed to work. I'm going off of https://kubernetes.io/docs/tasks/access-kubernetes-api/extend-api-custom-resource-definitions/#finalizers which says:

```
...Each controller then removes its finalizer from the list and issues the delete request again. This request only deletes the object if the list of finalizers is now empty, meaning all finalizers are done.
```

However, the behavior I was seeing was that if I issued a delete after removing the last/only finalizer, it would fail with object not found; the object appeared to be deleted once the last finalizer had been removed.